File: 2.Intermediate Level/6.Software-Design-Patterns/example/Structural/Adapter/Adapter.ts
interface Shape {
  getArea(): number;
}
interface AreaMeasurable {
  getAreaInCm(): number;
}
class Square implements Shape {
  private side: number;
  constructor(side: number) {
    this.side = side;
  }
  getArea(): number {
    return this.side * this.side;
  }
}
class Circle implements Shape {
  private radius: number;
  constructor(radius: number) {
    this.radius = radius;
  }
  getArea(): number {
    return Math.PI * this.radius * this.radius;
  }
}
// Adapter Class
class ShapeAreaCalculator implements AreaMeasurable {
  private shape: Shape;
  constructor(shape: Shape) {
    this.shape = shape;
  }
  getAreaInCm(): number {
    return this.shape.getArea() * 2.54; //This factor is used to convert the area from inches (assumed unit) to centimeters.
  }
}
// usage example :
const square: Square = new Square(5);
const circle: Circle = new Circle(3);
const squareAdapter: ShapeAreaCalculator = new ShapeAreaCalculator(square);
const circleAdapter: ShapeAreaCalculator = new ShapeAreaCalculator(circle);
console.log(`Area in cm : ${circleAdapter.getAreaInCm()}`);
console.log(`Area in cm : ${squareAdapter.getAreaInCm()}`);

File: 2.Intermediate Level/6.Software-Design-Patterns/example/Structural/Adapter/Adapter.py
from abc import ABC,abstractmethod
import math
class Shape(ABC):
  """Abstract class representing a shape"""
  @abstractmethod
  def getArea(self) -> int:
    """Abstract method to calculate the area of the shape"""
    pass

class AreaMeasurable(ABC):
  """Abstract class defining interface for getting area in cm"""
  @abstractmethod
  def getAreaInCm(self) -> int:
    """Abstract method to get the area in centimeters"""
    pass

class Square(Shape):
    def __init__(self,side):
        self.side =side
    def getArea(self) -> int:
        return self.side * self.side

class Circle(Shape):
    def __init__(self,radius):
        self.radius = radius
    def getArea(self) -> int:
        return math.pi * self.radius * self.radius

class ShapeAreaCalculator(AreaMeasurable):
    def __init__(self,shape):
        self.shape = shape
    def getAreaInCm(self) -> int:
        return self.shape.getArea() * 2.54 # This factor is used to convert the area from inches (assumed unit) to centimeters.

square = Square(5)
circle = Circle(3)
square_adapter = ShapeAreaCalculator(square)
circle_adapter = ShapeAreaCalculator(circle)

print(f'Area of square in cm: {square_adapter.getAreaInCm()}')
print(f'Area of circle in cm: {circle_adapter.getAreaInCm()}')